### PR TITLE
fix: `matchedText` module resolution incorrect length

### LINKED ISSUE #21636: Paths for module resolutions have ambiguous behaviour
**TypeScript Version:**  2.7.1

**Search Terms:** 
paths
module resolution

**Code**
```js
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@speedy/*/testing": ["src/*/dist"] 
      // the above is resolved to src/package-1/testing/dist
    }
  }
}
```

**Expected behavior:**
That the `*` in the resolved path matches returns only the matched part. Ex:  When having the following module `@speedy/package-1/testing`, the path is picked up by `@speedy/*/testing` => `@speedy/package-1/testing` and thus in the resolved part it should become`src/*/dist` => `src/package-1/dist`

**Actual behavior:**
The resolutions seems to be taking the `*`and the trailing part. Taking the same example as above`src/*/dist` => `src/package-1/testing/dist`.

I'd expect that the `/testing` after the star is omitted, the same as the current behavior for anything preceding the star which in fact is omitted.

### PR DESCRIPTION
**Pull request by alan-agius4**
`matchedText` was not properly extract the correct parts from the candidate.

Fixes #21636

### CODE REVIEW

**Comment by alan-agius4** [comment]
@rbuckton can you have a look at this please? Thanks

**Comment by alan-agius4** [comment]
@mhegazy can you have a look at this please? Thanks

**Comment by alan-agius4** [comment]
@andy-ms can you take a look at this please? Thanks

File: tests/cases/compiler/pathMappingBasedModuleResolution8_node.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+            "@speedy/*/testing": [
+               "*/dist/index.ts"
+            ]
+        }
+    }
+}
+
+// @filename: c:/root/index.ts
+import {x} from "@speedy/folder1/testing"
+declare function use(a: any): void;
```
Comment (andy-ms) [inline tests/cases/compiler/pathMappingBasedModuleResolution8_node.ts]: Are these two lines necessary? `--noUnusedLocals` is disabled by default.

File: src/compiler/core.ts [hunk 2]
```diff
@@ -2985,7 +2985,8 @@ namespace ts {
      */
     export function matchedText(pattern: Pattern, candidate: string): string {
         Debug.assert(isPatternMatch(pattern, candidate));
-        return candidate.substr(pattern.prefix.length, candidate.length - pattern.suffix.length);
+        const matchLength = candidate.length - pattern.suffix.length - pattern.prefix.length;
+        return candidate.substr(pattern.prefix.length, matchLength);
```
Comment (andy-ms) [inline src/compiler/core.ts]: Could use `slice` which takes start and end and is more common in the rest of the codebase. `substr` leads to nothing but problems (#20578).

Comment (alan-agius4) [reply src/compiler/core.ts]: I think you meant `substring` not slice `slice` but i get your point :)

File: tests/cases/compiler/pathMappingBasedModuleResolution8_node.ts (hunk 1 above)
Comment (alan-agius4) [reply tests/cases/compiler/pathMappingBasedModuleResolution8_node.ts]: In that case it's safe to remove

File: src/compiler/core.ts (hunk 2 above)
Comment (andy-ms) [reply src/compiler/core.ts]: Looks like we use them both about as often. According to [this](https://stackoverflow.com/questions/2243824/what-is-the-difference-between-string-slice-and-string-substring#2243835) `slice` and `substring` are basically equivalent with the exception of error handling -- if you avoid negative numbers, `NaN`, and `start > end`, they're the same.

**Comment by alan-agius4** [comment]
@andy-ms thanks for the review. Updated as suggested!

**Pull request merged**